applications/FluidDynamicsApplication/custom_elements/vms_adjoint_element.h

```diff
@@ -366,9 +366,115 @@ class KRATOS_API(FLUID_DYNAMICS_APPLICATION) VMSAdjointElement: public Element
+
+        noalias(rRightHandSideVector) = ZeroVector(TFluidLocalSize);
+
+        // Calculate RHS
+        // Calculate this element's geometric parameters
+        double gauss_weight;
+        array_1d<double, TNumNodes> N;
+        BoundedMatrix<double, TNumNodes, TDim> DN_DX;
+        GeometryUtils::CalculateGeometryData(this->GetGeometry(), DN_DX, N, gauss_weight);
+
+        // Calculate this element's fluid properties
+        double density;
+        array_1d<double, TDim> body_force;
+        FluidCalculationUtilities::EvaluateInPoint(this->GetGeometry(), N,
+                                                   std::tie(density, DENSITY),
+                                                   std::tie(body_force, BODY_FORCE));
+
+        double coeff = density * gauss_weight;
+
+        // Add the results to the velocity components (Local Dofs are vx, vy, [vz,] p for each node)
+        int local_index = 0;
+
+        for (unsigned int i_node = 0; i_node < TNumNodes; ++i_node) {
+            for (unsigned int d = 0; d < TDim; ++d) {
+                rRightHandSideVector[local_index++] +=
+                    coeff * N[i_node] * body_force[d];
+            }
+            ++local_index; // Skip pressure Dof
+        }
+
+        // For OSS: Add projection of residuals to RHS
+        if (rCurrentProcessInfo[OSS_SWITCH] == 1) {
+            KRATOS_ERROR
+                << "OSS projections are not yet supported with VMS adjoints.\n";
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    void CalculateLocalVelocityContribution(
+        MatrixType& rDampingMatrix,
+        VectorType& rRightHandSideVector,
+        const ProcessInfo& rCurrentProcessInfo) override
+    {
+        // Resize and set to zero the matrix
+        // Note that we don't clean the RHS because it will already contain body force (and stabilization) contributions
+        if (rDampingMatrix.size1() != TFluidLocalSize)
+            rDampingMatrix.resize(TFluidLocalSize, TFluidLocalSize, false);
+
+        noalias(rDampingMatrix) = ZeroMatrix(TFluidLocalSize, TFluidLocalSize);
+
+        const auto& r_geometry = this->GetGeometry();
+
+        // Get this element's geometric properties
+        double gauss_weight;
+        array_1d<double, TNumNodes> N;
+        BoundedMatrix<double, TNumNodes, TDim> DN_DX;
+        GeometryUtils::CalculateGeometryData(r_geometry, DN_DX, N, gauss_weight);
+
+        // Calculate this element's fluid properties
+        double density, viscosity;
+        array_1d<double, TDim> velocity, mesh_velocity, body_force;
+        FluidCalculationUtilities::EvaluateInPoint(
+            r_geometry, N, std::tie(density, DENSITY),
+            std::tie(velocity, VELOCITY),
+            std::tie(mesh_velocity, MESH_VELOCITY),
+            std::tie(viscosity, VISCOSITY),
+            std::tie(body_force, BODY_FORCE));
+
+        viscosity *= density;
+
+        const double element_size = this->CalculateElementSize(gauss_weight);
```

Comment:
Why not using  the  `ElementSizeCalculator`?

Reply:
It is to be consistent with the `VMS` element. That's the only reason.

Reply:
Considering that the same h formula is most probably implemented in `ElementSizeCalculator` I'd use the core utility. Also note that the `VMS` element is somehow (not officially) deprecated in favor of the `QSVSM` element.

Reply:
I would keep this since, it is more consistent with the actual `VMS` element. It is easier for undertsanding untill we retire it :). Would that be ok?